src/dc_imex.rs

```diff
@@ -448,36 +440,17 @@ pub unsafe fn dc_decrypt_setup_file(
     payload
 }
 
-pub unsafe fn dc_normalize_setup_code(
-    _context: &Context,
-    in_0: *const libc::c_char,
-) -> *mut libc::c_char {
-    if in_0.is_null() {
-        return ptr::null_mut();
-    }
+pub fn dc_normalize_setup_code(s: &str) -> String {
     let mut out = String::new();
-    let mut outlen;
-    let mut p1: *const libc::c_char = in_0;
-    while 0 != *p1 {
-        if *p1 as libc::c_int >= '0' as i32 && *p1 as libc::c_int <= '9' as i32 {
-            out += &format!("{}", *p1 as i32 as u8 as char);
-            outlen = out.len();
-            if outlen == 4
-                || outlen == 9
-                || outlen == 14
-                || outlen == 19
-                || outlen == 24
-                || outlen == 29
-                || outlen == 34
-                || outlen == 39
-            {
-                out += "-";
+    for p1 in s.chars() {
+        if p1 >= '0' && p1 <= '9' {
+            out += &format!("{}", p1);
```

Comment:
`write!(&mut out, "{}", p1)` is a bit more efficient I believe.

Reply:
Fixed with `push`